core/src/main/java/org/apache/calcite/rel/rules/ProjectMergeRule.java

```diff
@@ -68,6 +70,18 @@ public ProjectMergeRule(boolean force, ProjectFactory projectFactory) {
 
   //~ Methods ----------------------------------------------------------------
 
+  @Override public boolean matches(final RelOptRuleCall call) {
+    // We only merge logical project nodes with VolcanoPlanner. Since LogicalProject
+    // would be converted into physical nodes later. There's no need to fire this
+    // rule on physical project nodes again.
+    if (call.getPlanner() instanceof VolcanoPlanner
```

Comment:
I'm not in favor of having planner specific behavior inside the rules. I think creating two instances of the rule and passing them to different planners/rulesets is a cleaner solution.

Reply:
Yes, I agree with @zabetak.